File: matlab/lasso.m
%Not my own code.
% Function to run lasso by minimizing the following objective:
% f(beta) = 0.5*||y - X*beta||_2^2 + lambda*||beta||_1

% Inputs:
%   X: n-by-p matrix of input features
%   y: n-by-1 vector of output values
%   lambda: weight for l1 norm penalty

% Outputs:
%   beta: estimate of regression parameters

function beta = lasso(X,y,lambda)

% set options for optimization
miniter = 10;
maxiter = 4000;
tol = 1e-5;

% set fixed step size
t = 0.0001;

% initialize beta
p = size(X,2);
beta = zeros(p,1);

Grad = zeros(maxiter, p);

% optimize lasso objective 
for iter = 1:maxiter
    % compute gradient of f(beta) with respect to beta
    grad = (-(y - X*beta)'*(X))';
    Grad(iter, :) = grad;
    % compute new estimate of beta by appling soft thresholding operator
    beta_new = softthresh(beta-(t*grad),t*lambda);
     % compute new objective value
    obj_new = sum((y-X*beta_new).^2)/2 + lambda*sum(abs(beta_new));
    % check convergence
    if (iter >= miniter && abs(obj_new-obj) < tol)
        break;
    end
    % store new variables
    beta = beta_new;
    obj = obj_new;
end

%Grad
end

% Soft thresholding
function alpha = softthresh(beta,eta)
        
    beta(find(beta>eta)) = beta(find(beta>eta)) - eta;
    beta(find(beta<-eta)) = beta(find(beta<-eta)) + eta;
    beta(find(beta >= -eta & beta <= eta)) = 0;
    
    alpha = beta;
    %%% alpha = ?
    %%% Fill in computation of soft thresholding operator
    %%% Elements of beta whose values are larger than eta are decremented by eta (subtract eta)
    %%% Elements of beta whose values are smaller than -eta are incremented by eta (add eta)
    %%% Elements of beta whose value is in the range [-eta,eta] are set exactly to zero
    
end

File: cpp/src/lasso.cpp
/*
 * lasso.cpp
 *
 *  Created on: Nov 23, 2014
 *      Author: manu
 */

#include <iostream>
#include <Eigen/Dense>
using namespace Eigen;
using namespace std;


void softThreshold(VectorXd* beta, double eta){
	for (int i=0;i<beta->size() ;i++){
		if(beta->operator()(i) < -1 *eta){
			beta->operator()(i)+=eta;
		}
		else if (beta->operator()(i) > eta){
			beta->operator()(i)-=eta;
		}
		else{
			beta->operator()(i) = 0;
		}
	}
}


VectorXd lasso(MatrixXd X, VectorXd Y, double lambda){

	int minIter = 10;
	int maxIter = 4000;
	double epsilon = pow(10,-5);

	//step size = t;
	double t = 0.0001;

	int size = X.cols();
	VectorXd beta = VectorXd::Zero(size);
	MatrixXd Grad = MatrixXd::Zero(maxIter, size);
	double obj = 0.0;
	for(int iter=0; iter<maxIter ; iter++){
		MatrixXd grad = X.transpose() * ((X*beta) - Y);
		VectorXd beta_new = beta - t*grad;
		softThreshold(&beta_new, t*lambda);
		double obj_new = ((Y-(X*beta_new)).squaredNorm())/2  - (lambda * (beta_new.cwiseAbs().sum()));
		if(iter > minIter  && (abs(obj_new - obj) < epsilon)){
			break;
		}
		beta = beta_new;
		obj = obj_new;
	}
	return beta;
}



//int main()
//{
////	MatrixXd X(5,5);
////	X << 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25;
////	cout << X << endl;
////	VectorXd Y(5);
////	Y<< 1,-1,1,1,-1;
////	cout << Y << endl;
////
////	int size = X.cols();
////	VectorXd beta = VectorXd::Zero(size);
////	MatrixXd Grad = MatrixXd::Zero(100, size);
////	double obj = 0.0;
////
////	cout << beta << endl;
////	MatrixXd grad = X.transpose() * ((X*beta) - Y);
////
////	cout <<"grad" << endl<< grad << endl;
////	double t= 0.001;
////	VectorXd beta_new = beta - t*grad;
////	double lambda = 0.01;
////	cout << "beta_new" << endl<< beta_new << endl;
////	softThreshold(&beta_new, t*lambda);
////
////	cout << "beta_new" << endl<< beta_new << endl;
////
////
////
////	cout << "objective" << endl << obj_new << endl;
//
//
//}





